core/src/main/scala/org/apache/spark/crypto/CryptoInputStream.scala

```diff
@@ -0,0 +1,439 @@
+      resetStreamOffset(pos)
+      skipped
+    }
+  }
+
+  /** ByteBuffer read. */
+  override def read(buf: ByteBuffer): Int = {
+    checkStream
+    if (isReadableByteChannel) {
+      val unread: Int = outBuffer.remaining
+      if (unread > 0) {
+        val toRead: Int = buf.remaining
+        if (toRead <= unread) {
+          val limit: Int = outBuffer.limit
+          outBuffer.limit(outBuffer.position + toRead)
+          buf.put(outBuffer)
+          outBuffer.limit(limit)
+          toRead
+        }
+        else {
+          buf.put(outBuffer)
+        }
+      }
+      val pos: Int = buf.position
+      val n: Int = (in.asInstanceOf[ReadableByteChannel]).read(buf)
+      if (n > 0) {
+        streamOffset += n
+        decrypt(buf, n, pos)
+      }
+      if (n >= 0) {
+        unread + n
+      }
+      else {
+        if (unread == 0) {
+          -1
+        }
+        else {
+          unread
+        }
+      }
+    }
+    else {
+      var n: Int = 0
+      if (buf.hasArray) {
+        n = read(buf.array, buf.position, buf.remaining)
+        if (n > 0) {
+          buf.position(buf.position + n)
+        }
+      }
+      else {
+        val tmp: Array[Byte] = new Array[Byte](buf.remaining)
+        n = read(tmp)
+        if (n > 0) {
+          buf.put(tmp, 0, n)
+        }
+      }
+      n
+    }
+  }
+
+  /**
+   * Decrypt all data in buf: total n bytes from given start position.
+   * Output is also buf and same start position.
+   * buf.position() and buf.limit() should be unchanged after decryption.
+   */
+  def decrypt(buf: ByteBuffer, n: Int, start: Int) {
+    val pos: Int = buf.position
+    val limit: Int = buf.limit
+    var len: Int = 0
+    while (len < n) {
+      buf.position(start + len)
+      buf.limit(start + len + Math.min(n - len, inBuffer.remaining))
+      inBuffer.put(buf)
+      try {
+        decrypt(decryptor, inBuffer, outBuffer, padding)
+        buf.position(start + len)
+        buf.limit(limit)
+        len += outBuffer.remaining
+        buf.put(outBuffer)
+      }
+      finally {
+        padding = afterDecryption(decryptor, inBuffer, streamOffset - (n - len), iv)
+      }
+    }
+    buf.position(pos)
+  }
+
+  override def available: Int = {
+    checkStream
+    in.available + outBuffer.remaining
+  }
+
+  override def markSupported: Boolean = {
+    false
+  }
+
+  override def mark(readLimit: Int) {
+  }
+
+  override def reset() {
+    throw new IOException("Mark/reset not supported")
+  }
+
+  override def read: Int = {
+    if ((read(oneByteBuf, 0, 1) == -1)) -1 else (oneByteBuf(0) & 0xff)
+  }
+
+  def checkStream() {
+    if (closed) {
+      throw new IOException("Stream closed")
+    }
+  }
+
+  /** Forcibly free the direct buffers. */
+  def freeBuffers {
+    CryptoStreamUtils.freeDB(inBuffer)
+    CryptoStreamUtils.freeDB(outBuffer)
+    cleanBufferPool
+  }
+
+  /** Clean direct buffer pool */
+  def cleanBufferPool {
+    var buf: ByteBuffer = null
+    while ((({
```

Comment:
This looks like a lot of parentheses.

Can't you use `bufferPool.asScala.foreach(CryptoStreamUtils.freeDB)` instead?